test/gurka/test_reach.cc

```diff
@@ -0,0 +1,233 @@
+std::shared_ptr<baldr::GraphReader> TestReach::reader;
+
+TEST_F(TestReach, ReachWithNoClosures) {
+  // Create costing
+  // - type: auto
+  // - flow mask: "default" (includes current) to consider live speeds
+  // - filter_closures: disable filter closed edges
+  sif::CostFactory factory;
+  CostingOptions co;
+  co.set_costing(valhalla::auto_);
+  co.set_flow_mask(kDefaultFlowMask);
+  co.set_filter_closures(false);
+
+  auto costing = factory.Create(co);
+  auto edge = gurka::findEdgeByNodes(*reader, closure_map.nodes, "A", "B");
+
+  // check its reach
+  loki::Reach reach_checker;
+  auto reach = reach_checker(std::get<1>(edge), std::get<0>(edge), 50, *reader, costing);
+
+  // all edges should have the same in/outbound reach
+  EXPECT_EQ(reach.inbound, reach.outbound);
+  EXPECT_EQ(reach.inbound, 9);
+  EXPECT_EQ(reach.outbound, 9);
+}
+
+TEST_F(TestReach, ReachWithClosures) {
+  LiveTrafficCustomize close_edge = [](baldr::GraphReader& reader, baldr::TrafficTile& tile,
+                                       uint32_t index, baldr::TrafficSpeed* current) -> void {
+    close_bidir_edge(reader, tile, index, current, "AB", closure_map);
+    close_bidir_edge(reader, tile, index, current, "BC", closure_map);
+    close_bidir_edge(reader, tile, index, current, "AD", closure_map);
+    close_bidir_edge(reader, tile, index, current, "BE", closure_map);
+  };
+  test::customize_live_traffic_data(closure_map.config, close_edge);
+
+  // Create costing
+  // - type: auto
+  // - flow mask: "default" (includes current) to consider live speeds
+  // - filter_closures: enabled, to remove closed edges
+  sif::CostFactory factory;
+  loki::Reach reach_checker;
+
+  CostingOptions co;
+  co.set_costing(valhalla::auto_);
+  co.set_flow_mask(kDefaultFlowMask);
+  co.set_filter_closures(true);
+  // Check reach for AB - enclosed by closures, so reach should be 0
+  {
+    auto costing = factory.Create(co);
+    auto edge = gurka::findEdgeByNodes(*reader, closure_map.nodes, "A", "B");
+    auto reach = reach_checker(std::get<1>(edge), std::get<0>(edge), 50, *reader, costing);
+
+    EXPECT_EQ(reach.inbound, reach.outbound);
+    EXPECT_EQ(reach.inbound, 0);
+    EXPECT_EQ(reach.outbound, 0);
+  }
+  // Check reach for HI - open, but reduced reach without ignoring closures
+  {
+    auto costing = factory.Create(co);
+    auto edge = gurka::findEdgeByNodes(*reader, closure_map.nodes, "H", "I");
+    auto reach = reach_checker(std::get<1>(edge), std::get<0>(edge), 50, *reader, costing);
+
+    EXPECT_EQ(reach.inbound, 7);
+    EXPECT_EQ(reach.outbound, 7);
+  }
+
+  // Check reach for AB ignoring closures
+  {
+    co.set_filter_closures(false);
+    auto costing = factory.Create(co);
+    auto edge = gurka::findEdgeByNodes(*reader, closure_map.nodes, "A", "B");
+    auto reach = reach_checker(std::get<1>(edge), std::get<0>(edge), 50, *reader, costing);
+
+    EXPECT_EQ(reach.inbound, 8);
```

Comment:
The in/outbound reach should be `9` (the same as with no closures). However what I'm observing is that edges leaving `A` are not looked at when expanding in forward & reverse directions. I have not yet figured why the expansion is omitting these edges. Let me know if you an idea as to why it would be doing that.

Reply:
if u-turns are disallowed, i could see how that would happen. consider the forward direction of the reach check. it would start in the closure get out of it but not be able to get back in. in this case it wouldnt be able to reach the A node. but that only works if its not able to make a uturn at B, stay in the closure and reach A. to visualize this we could modify the reach algorithm to print out the node lat lons as they are visited just to be sure. or we could just do a one time debug session to watch what nodes are being seen when.

Reply:
Okay, so this is what I figured is going on: When we're expanding forward (similar thing happens for reverse expansion too), at some some point we go from consecutive closed edge(s) to an open edge, eg AB -> BE , which enables closure pruning. Now when we expand further from BE, one path goes to DE (another open edge), and then DE tries expanding to AD which is closed. Since pruning is on we never expand from DE into AD which explains why the reach is off by 1. CMIIW.